src/protocols/kubernetes/settings.c

```diff
@@ -25,6 +25,14 @@
 
 /* Client plugin arguments */
 const char* GUAC_KUBERNETES_CLIENT_ARGS[] = {
+    /**
+     * This first argument defines the protocol version in use so that the
+     * client knows how to handle talking to different versions of guacd.
+     * If this is omitted the client may choose not to enable certain
+     * features.
+     */
+    "VERSION_1_1_0",
```

Comment:
Thoughts on handling this automatically within `guac_protocol_send_args()`?

Reply:
Fine with me, I'll take a look...

Reply:
Okay,  I think I've managed to accomplish this.  Only odd thing at this point is that, when both client and server are updated, I see the following message in the server log:

    guacd[35020]: DEBUG:	Parameter "(null)" omitted. Using default value of "America/New_York".

Is there something else I need to do on the receiving side to pull this off correctly so that it doesn't end up as a `null` parameter?  It seems to work okay - the timezone is getting set - but the message indicates something isn't quite right.

Reply:
For a call like:

```c
guac_user_parse_args_string(user, GUAC_RDP_CLIENT_ARGS, argv,
         IDX_TIMEZONE, user->info.timezone);
```

With the new timezone parameter being at the end, the `IDX_TIMEZONE` is likely off by one, resulting in the `NULL` value at the end of `GUAC_*_CLIENT_ARGS` being pulled in for logging rather than the actual name of the parameter. This is probably due to the addition of `IDX_PROTOCOL_VERSION`.

Just as `guac_protocol_send_args()` has been modified to automatically include the protocol version, I suggest also automatically parsing out the protocol version within the code handling `connect`, storing the value somewhere within the `guac_user` just as we do the other parts of the handshake. This would avoid the need for protocol implementations to be aware of this detail.

Could be as simple as:

* Verifying receipt of the expected number of arguments
* Storing `argv[0]` somewhere sensible in `guac_user`
* Using `argc - 1` and `argv + 1` for the call to `guac_client_add_user()`

Existing protocol implementations would then need zero modification, except those that wish to check the protocol version.

Reply:
> Could be as simple as:
> Verifying receipt of the expected number of arguments
> Storing argv[0] somewhere sensible in guac_user
> Using argc - 1 and argv + 1 for the call to guac_client_add_user()

Is there a way to do this outside of the protocol-specific parameter parser?  The expected number of arguments will vary protocol-to-protocol, so is there a way to know prior to the `guac_client_add_user()` call what the `parser->argc` value should be, and whether or not to subtract from `argc` and do `argv + 1`?

Reply:
> Is there a way to do this outside of the protocol-specific parameter parser?

Yes. There's no protocol specific parser - the Guacamole protocol requires that the `connect` instruction include a value for each argument declared in `args`, in the same order as specified in `args`.

Regardless of the protocol implementation, the Guacamole protocol would guarantee one of the following has to be true:

1. The client side is an up-to-date implementation and will populate that initial `VERSION_1_1_0` argument with its own protocol version.
2. The client side is older and will interpret `VERSION_1_1_0` as the literal name of an argument. As that argument presumably does not exist and does not have a value, it will be blank in the reponse.
3. The client does not correctly implement the Guacamole protocol and the argument count does not match (the server should close the connection at this point as all bets are off).

Reply:
Side note: even if it proves inconvenient to fully verify the argument count outside the protocol implementation (doing so would require iterating through the args until we find the `NULL`), we _do_ know that any conforming Guacamole implementation will provide at least one argument value because `guac_protocol_send_args()` will now always send at least one argument value (the version), so lacking at least one value within the received `args` would be a protocol violation.

We could, of course, choose to be tolerant of that, leave total rejection up to the protocol implementation, and handle the lack of even one argument value as equivalent to leaving that initial value blank.

Reply:
Okay, so I've played around with this a bit more and got it to the point where it works across various combinations of server and client versions in a consistent and expected way.  As it is right now, I ended up just removing the first argument on the guacd side.  The client implementation does not currently populate a value for the protocol version, it simply examines the protocol that the server sends and adjusts accordingly (see apache/guacamole-client#348).

The various scenarios are:
* Server 1.1.0, Client 1.1.0: Server sends protocol version and args to client, client examines and sets protocol version, adjusts handshake, and sends back an empty value for the protocol version argument.  The server then ignores this argument when joining the user to the connection by decrementing `argc` and passing through `argv + 1` to the join function.
* Server 1.1.0, Client 1.0.0 (or lower): Server sends protocol version and args to client, client has no knowledge of protocol version, so does not adjust handshake at all, and sends back an empty value for the protocol version argument.  The server ignores/strips the argument when joining using the same method as above.
* Server 1.0.0 (or lower), Client 1.1.0: Server sends arguments to client, without any protocol version.  The client checks for protocol version and does not see it, so it does not add any of the additional protocol-related functionality and simple passes through the supported handshake and arguments.  The server processes all of the arguments as usual.

So, my two questions would be:
* Is there some use-case I'm missing here where checking the number of provided/expected arguments in the handshake actually helps anything out?  As it is, it seems to work with various combinations of the server and client, so I'm not sure this is valuable?
* I currently have the client where it simply sends an empty argument back for the protocol version - is there any reason to have it populate something else in that argument, like the actual protocol version it's using?  It doesn't seem to need it right now in order to function properly, but maybe it's worth just implementing it this way, anyway, to future-proof it?  Or just leave it as-is?

Reply:
> I currently have the client where it simply sends an empty argument back for the protocol version

I went back and reworked this on the client-side a little bit, and I'm now sending back the negotiated protocol version from client to server as an argument.  I'm still not doing anything with it on the server-side, but I'm no longer just passing an empty argument.